### PR TITLE
COM-2726: Added B2B Account Hierarchy for B2B Buyers.

### PR DESCRIPTION
**Pull request by sagarraut0007**
Developed B2B Account's Hierarchy tree to show all accounts in tree format and also the tree nodes have action menus to "View Account" and "Change Parent Account".
Action Menus only shown to user who have Admin or Purchaser roles.
This tree will only be shown to B2B buyer.

### CODE REVIEW

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                //set menu's on all current accounts children accounts.
+                item.isViewAccount = self.isUserAdmin() || self.isUserPurchaser();
+                item.isChangeParentAccount = self.isUserAdmin();
+            }
+
+            item.account = self.getAccount(item.id, accounts);
+
+            if (item.children) {
+                //do recursive call for children's 
+                self.processAccountHierarchyChildrens(item.children, accounts,
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: You can replace this call with a for loop and call itself.
```
for (var i = 0; i < item.children.length; i++) {
    self.setAccountOnHierarchy(item.children[i], accounts, isDescendant);
}
```

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            return this.isUserHavingBehavior(1000);
+        },
+        isUserPurchaser: function () {
+            //purchaser have 1005 behavior
+            return !this.isUserAdmin() && this.isUserHavingBehavior(1005);
+        },
+        processAccountHierarchy: function (data) {
+            var self = this;
+            if (data.hierarchy) {
+                self.setAccountOnHierarchy(data.hierarchy, data.accounts);
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: Optional but could explicitly pass in false here for the 3rd param.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            }
+            return data.hierarchy;
+        },
+        setAccountOnHierarchy: function (item, accounts, isCurrentAccountChildrens) {
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: I would personally rename the `isCurrentAccountChildrens` param to `isDescendantOfAssociatedAccount` or maybe more simply `isDescendant` with a comment in the code. I see you use it to control behavior during recursion.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        initialize: function () {
+            var self = this;
+            self.initApiModel();
+            self.apiGet({ id: require.mozuData('user').accountId }).then(function (response) {
+                self.set("hierarchy", self.processAccountHierarchy(response.data));
+                self.syncApiModel();
+            });
+            self.set("isUserAdmin", self.isUserAdmin());
+        },
+        isUserHavingBehavior: function (behaviorId) {
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: Name `userHasBehavior` or `doesUserHaveBehavior` makes more sense.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            else {
+                //set menu's on all current accounts children accounts.
+                item.isViewAccount = self.isUserAdmin() || self.isUserPurchaser();
+                item.isChangeParentAccount = self.isUserAdmin();
+            }
+
+            item.account = self.getAccount(item.id, accounts);
+
+            if (item.children) {
+                //do recursive call for children's
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: To be more accurate the terminology "childrens" or "children's" should be descendant. That's the terminology I'm using for all the hierarchy/tree design. Conversely parents are ancestors.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            return data.hierarchy;
+        },
+        setAccountOnHierarchy: function (item, accounts, isCurrentAccountChildrens) {
+            var self = this;
+            // isCurrentAccountChildrens will be null/false initially but when current account is logged in users account then set it to true. so we can use this to show menu on children.
+            if (!isCurrentAccountChildrens) {
+                //Check whether this is a current logged in users associated account.
+                isCurrentAccountChildrens = require.mozuData('user').accountId === item.id;
+                //set "View Account" menu on current account.
+                item.isViewAccount = isCurrentAccountChildrens && (self.isUserAdmin() || self.isUserPurchaser());
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: Logic is fine but the naming `isViewAccount` should be `canViewAccount` or `enableViewAccount`. Same for `isChangeParentAccount`.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        processAccountHierarchy: function (data) {
+            var self = this;
+            if (data.hierarchy) {
+                self.setAccountOnHierarchy(data.hierarchy, data.accounts);
+            }
+            return data.hierarchy;
+        },
+        setAccountOnHierarchy: function (item, accounts, isCurrentAccountChildrens) {
+            var self = this;
+            // isCurrentAccountChildrens will be null/false initially but when current account is logged in users account then set it to true. so we can use this to show menu on children.
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: Comment is good but grammar. I'd recommend something like below. Food for thought.
```
// This parameter will be false until recursion finds the account associated
// with the current user. On further recursion it will enable descendant behaviors.
```

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            item.account = self.getAccount(item.id, accounts);
+
+            if (item.children) {
+                //do recursive call for children's 
+                self.processAccountHierarchyChildrens(item.children, accounts,
+                    isCurrentAccountChildrens
+                );
+            }
+        },
+        processAccountHierarchyChildrens: function (childrens, accounts, isCurrentAccountChildrens) {
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: This first if will never be hit as you never call this method with the root node. You can delete this method and get this to work with just the `setAccountOnHierarchy` method. See other comment.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        },
+        isUserHavingBehavior: function (behaviorId) {
+            var behaviors = require.mozuData('user').behaviors;
+            if (behaviors) {
+                return behaviors.includes(behaviorId);
+            }
+            return false;
+        },
+        isUserAdmin: function () {
+            //Only admin should have 1000 behavior.
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js:82]: Can you add the behavior name to the comment? What is 1000? Same with the others.

File: scripts/modules/b2b-account/account-hierarchy.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            $(document).ready(function () {
+                //close menu if we click on outside of menu
+                $(document).on("click", function (event) {
+                    if (!$(event.target).closest(".dropdown-content").length && !$(event.target).closest(".three-dots").length) {
+                        $(".account-hierarchy .dropdown-content").removeClass("active");
+                    }
+                });
+
+                //fix margins
+                $(".node:not(:has(.caret))").css("margin-left", "-10px");
```
Comment (ghost) [inline scripts/modules/b2b-account/account-hierarchy.js]: Minor but ideally you would add/remove a class from the HTML element instead of adding CSS directly. I'd recommend adding a `.fix-margin` class or something with the rule and then use the jQuery method `addClass(name)`.

**Review by ghost — APPROVED** [review]
Overall looks really good. Requested some minor changes.

File: scripts/modules/b2b-account/account-hierarchy.js (hunk 1 above)
Comment (sagarraut0007) [reply scripts/modules/b2b-account/account-hierarchy.js:110]: I would prefer to keep it as is by using processAccountHierarchyChildrens() method because that method has another logic as well, like identify the top-level node of the tree and pass either a single object or list of children's to setAccountOnHierarchy() method.

**Comment by sagarraut0007** [comment]
@michaelsadlonkibo , Resolved all the issues. please take a look again?

**Review by ghost — APPROVED** [review]

File: stylesheets/modules/b2b-account/account-hierarchy.less [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        left: -8px;
+        position: relative;
+        top: 1.5px;
+    }
+
+    .caret::after {
+        border-left: 2px solid rgb(100, 100, 100);
+    }
+
+    .caret-user {
```
Comment (ghost) [inline stylesheets/modules/b2b-account/account-hierarchy.less]: Caret user isn't used anymore. Can remove it as well as the unused image files.

File: labels/de-DE.json [hunk 12]
```diff
@@ -289,8 +293,8 @@
   "mobileOrderConfirmationMessage": "{0} Bestellbenachrichtigungen. Wir werden Ihnen eine SMS senden, wenn sich Ihr Bestellstatus ändert. Die Meldungsfrequenz kann variieren. Antworte {1}, um dich abzumelden. Es können Nachrichten- und Datentarife gelten",
   "mobileOrderOptInMessage": "Vielen Dank, dass Sie sich für SMS-Benachrichtigungen für Ihre Curbside-Bestellung angemeldet haben. Wir werden Sie benachrichtigen, wenn Ihre Bestellung zur Abholung bereit ist.",
   "mobilePickLocation": "Ramasser / Expédier de:",
-  "mobilePickupReminder":"Erinnerung - Artikel in Ihrer Bestellung {0} können abgeholt werden. Überprüfen Sie die Bestellung unter {1}",
-  "mobileReadyForPickup":"Artikel in Ihrer Bestellung {0} können um {1} abgeholt werden. Überprüfen Sie die Bestellung unter {2}",
+  "mobilePickupReminder": "Erinnerung - Artikel in Ihrer Bestellung {0} können abgeholt werden. Überprüfen Sie die Bestellung unter {1}",
```
Comment (ghost) [inline labels/de-DE.json]: Revert unrelated whitespace changes in both locale files. This repo is public so should minimize changes. If you have format on save enabled please disable it.

**Review by ghost — APPROVED** [review]
Looks fine with the latest fixes but need to make some small changes. See my comments. Make sure to remove the unused image files.

In the future if we make fixes after code review don't squash all the commits together. It's really hard for me to see what changed. You can squash after review is done.

File: stylesheets/modules/b2b-account/account-hierarchy.less (hunk 11 above)
Comment (SharmaBhushan) [reply stylesheets/modules/b2b-account/account-hierarchy.less]: Done

**Review by ghost — APPROVED** [review]
